keps/sig-api-machinery/4590-schema-cel-variables/README.md

```diff
@@ -0,0 +1,1115 @@
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+<!--
+This section is incredibly important for producing high-quality, user-focused
+documentation such as release notes or a development roadmap. It should be
+possible to collect this information before implementation begins, in order to
+avoid requiring implementors to split their attention between writing release
+notes and implementing the feature itself. KEP editors and SIG Docs
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+CRD Validation Rules were added in Kubernetes X.XX. It was a huge leap in
+terms of expressibility of validation rules in Kubernetes schemas, but is missing
+the ability to define common variables for use in multiple expressions - leading
+to verbose and unperformant validation rules.
+
+This was recognized during development of ValidatingAdmissionPolicy which also uses
+CEL rules. ValidatingAdmissionPolicy added the ability to define variables to
+refer within the policy rules. This PR aims to accomplish the same thing for
+CRDs and Native Type CEL expressions in schemas expressed with OpenAPI.
+
+## Motivation
+
+Improve readability, performance, maintainability of common conditions used
+in multiple locations in CEL validations.
+
+### Goals
+
+* Ability to add variables in Schema that can be referred to by multiple CEL validation rules
+* Ability to refer to variables defined in parent schemas, with proper shadowing rules
+* Ability to refer to previously defined variables from within other variable definitions
+* Ability to refer to oldSelf in variables (either through implicit optionalOldSelf, or opt-in like normal rules. implicit optionalOldSelf would ahve simpler rules since same set of vars are used for all types of rules)
+* Backwards compatibility with current schemas
+
+### Non-Goals
+
+* Ability to Embed Parameterized Schemas by overriding variables  - This is seen as a future work
+
+## Proposal
+
+The proposal is to add a new `x-` extension to our schema validation support code:
+`x-kubernetes-validation-variables` with the following format:
+
+```go
+type XValidationVariables []ValidationVariable
+type ValidationVariale {
+  Name string `json:"name"`
+  Expression string `json:"expression"`
+}
+```
+
+The schema validator will allow shadowing of variables defined in parent 
+schemas.
+
```

Comment:
Explain variable scope and shadowing in more detail?  My assumption is that a variable in the data tree is in scope for all children (unless shadowed)